### ISSUE TITLE
Class Tree branch

### ISSUE THREAD

**Issue by schriftgestalt** [googlei18n/glyphsLib#136]
I just pushed my attempt to implement proper classes instead of using a dict to represent the data in the .glyphs file. 

I did already a bit more as I added some logic that makes it behave a bit more than the GlyphApp python API. One use case of this is that one could use the builder inside of Glyphs.

It needs a lot of cleanup, still ;)

Can you please have a look at and comment on the changes? Thanks

**Comment by anthrotype** [comment]
Thanks Georg for working on this. It's a huge patch, it'll take some time to review! :)

>  One use case of this is that one could use the builder inside of Glyphs.

This sounds interesting. Can you provide an example of that would work?

**Comment by schriftgestalt** [comment]
I tried to split it into smaller commits...

One could, instead of importing 'classes.py' rely on the apps API. Import the builder and use it to write an UFO, or variation font.

**Comment by m4rc1e** [comment]
Thanks Georg!,

I love this. 'll be porting some of my scripts so I can run them from the cmdline. 

I'll add some issues if I come across them.

**Comment by m4rc1e** [comment]
When installing this branch as a dependency for fontmake. We get this traceback when we execute [run.sh](https://github.com/googlei18n/fontmake/blob/master/test/run.sh)

```

INFO:fontmake.font_project:Saving TTF for MyFont-Regular
WARNING:ufo2ft.fontInfoData:Underline thickness not set in UFO, defaulting to UPM * 0.05
WARNING:ufo2ft.fontInfoData:Underline position not set in UFO, defaulting to UPM * -0.075
WARNING:ufo2ft.fontInfoData:Underline thickness not set in UFO, defaulting to UPM * 0.05
INFO:fontmake.font_project:Building master UFOs and designspace from Glyphs source
INFO:glyphsLib:Parsing .glyphs file
____loads
INFO:glyphsLib:Casting parsed values
INFO:glyphsLib:Loading to UFOs
INFO:glyphsLib.util:Writing master_ufo/GuidelineTest-.ufo
Traceback (most recent call last):
  File "/Users/marc/Documents/fontmake/env/bin/fontmake", line 9, in <module>
    load_entry_point('fontmake==1.2.0.dev0', 'console_scripts', 'fontmake')()
  File "build/bdist.macosx-10.11-intel/egg/fontmake/__main__.py", line 171, in main
  File "build/bdist.macosx-10.11-intel/egg/fontmake/font_project.py", line 341, in run_from_glyphs
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
    return func(*args, **kwds)
  File "build/bdist.macosx-10.11-intel/egg/fontmake/font_project.py", line 71, in build_master_ufos
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/glyphsLib/__init__.py", line 94, in build_masters
    ufos, master_dir, designspace_instance_dir, instance_data)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/glyphsLib/interpolation.py", line 65, in build_designspace
    write_ufo(font, master_dir)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/glyphsLib/util.py", line 40, in write_ufo
    ufo.save(out_path)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/defcon/objects/font.py", line 736, in save
    self.layers.save(writer, saveAs=saveAs, progressBar=progressBar)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/defcon/objects/layerSet.py", line 317, in save
    layer.save(glyphSet, saveAs=saveAs, progressBar=progressBar)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/defcon/objects/layer.py", line 612, in save
    self.saveGlyph(glyph, glyphSet, saveAs=saveAs)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/defcon/objects/layer.py", line 628, in saveGlyph
    glyphSet.writeGlyph(glyph.name, glyph, glyph.drawPoints)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/ufoLib/glifLib.py", line 345, in writeGlyph
    data = writeGlyphToString(glyphName, glyphObject, drawPointsFunc, formatVersion=formatVersion)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/ufoLib/glifLib.py", line 582, in writeGlyphToString
    _writeLib(glyphObject, writer)
  File "/Users/marc/Documents/fontmake/env/lib/python2.7/site-packages/ufoLib/glifLib.py", line 742, in _writeLib
    plistWriter.writeValue(lib)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 252, in writeValue
    self.writeDict(value)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 281, in writeDict
    self.writeValue(value)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 258, in writeValue
    self.writeArray(value)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 287, in writeArray
    self.writeValue(value)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 252, in writeValue
    self.writeDict(value)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 281, in writeDict
    self.writeValue(value)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plistlib.py", line 260, in writeValue
    raise TypeError("unsuported type: %s" % type(value))
TypeError: unsuported type: <type 'NoneType'>
GuidelineTest failed to build
```

**Comment by schriftgestalt** [comment]
I didn’t run the full tests, yet. I’ll have a look, soon.

This is what I tested with. 
```
path = "/Users/georg/Stuff/filename"
f = open(path + ".glyphs")
font = glyphsLib.load_to_ufos(f)[0]
font.save(path + ".ufo")
```

**Comment by moyogo** [comment]
@schriftgestalt I started cleaning up in https://github.com/googlei18n/glyphslib/compare/Class_tree_new...moyogo:classtree-cleanup. Now the tests pass but there are still things to fix.

**Comment by schriftgestalt** [comment]
I added some comments but could you send a pull request? Thanks.

**Comment by anthrotype** [comment]
I've spent all day reviewing the code in `Class_tree_new` branch. I have to say this doesn't look like a patch to me, but rather a complete rewrite. I find quite hard even to start discussing about it... The changes are simply *too many*.

The problem is not simply to make the few tests pass. It's about designing an interface which we're comfortable working with. And implementing that in a way that we can maintain in the future, instead of rushing things out half-baked and accumulating technical debt.

Let's step back for a moment and think about what we are trying to do here. I believe Georg's end goal is to make something like a 'defcon' for Glyphs.app, e.g. a set of Python objects to work with .glyphs source data, which can be used independently from the Glyphs.app, but still follows a similar API.

What we currently have in glyphsLib/master is instead -- to stay within the same UFO metaphors -- just the "ufoLib" part of the puzzle: i.e., a way to parse a .glyphs source's content in the form of raw python data structures (dict, list, tuple, int, float, etc.), which we then use (in builder.py) to directly populate a defcon.Font instances. All in all, this has worked well so far, since most of the libraries used by fontmake are UFO-based. This approach also has the advantage that the same code path can be used to produce binary fonts from either Glyphs or UFO source -- of course, provided the .glyphs data is first converted to UFO by glyphsLib.

Now, even if we had a fully fledged API to manipulate glyphs sources outside of Glyphs.app that is comparable to defcon for UFOs, I believe we would still like to _not_ split up the pipeline into `ufo2ft`, on the one hand, and a "glyphs2ft", on the other. If anything, what we would like to see is an even more abstract interface for font source data (fontParts?) -- but that's outside of glyphsLib's scope.

This, if fontmake is happy to stick with defcon-based objects for a while, what is the real advantage of an object-oriented interface to .glyphs sources, beyond the syntactic sugar of using instance attributes (i.e. `font.glyphs`) instead of dictionaries (`font["glyphs"]`)? I do understand users of Glyphs.app would like to be able to run their own scripts outside of the app, with few or less modifications. Georg also suggested the possibility of using glyphsLib to convert native `GlyphsApp` objects to UFOs when running from within the app. This is all very cool!
But, in order to achieve that I think we should not risk to compromise the consistency, maintainability, <insert buzz word here>, of the current glyphsLib code.

And I wonder if this goal could be better achieved by separating the concerns of parsing/writing syntactically correct .glyphs file (i.e. currently glyphsLib more or less currently does), with that of providing an higher-level object-oriented interface to manipulate the raw data (again, something along the lines of ufoLib vs defcon). The latter may well live in a module or sub-package inside glyphsLib; however, it would build upon the current low-level parser and casting modules, rather than (partly and somewhat confusingly) replacing them, as in the `Class_tree_new` branch.

I'm really sorry if this looks like I'm dismissing the huge effort Georg has put into all this; I really do appreciate it!!!
But perhaps the best thing to do is wait when we all meet face-to-face in a few weeks in Berlin, where hopefully we can sit down and hack this together.

**Comment by schriftgestalt** [comment]
I understand your concern. But if you look at the builder, the code can be much cleaner already. 

The code in classes.py needs some clean up and my coding style might be a bit unpytonic. And there are some workaround that will be removed once the wrapper is finished (proper getter/setter for all properties).

Lets have a look at it together.

**Comment by moyogo** [comment]
@schriftgestalt I started the 'classtree' branch, which is rebased. I’ll try to finish it when I have time but feel free to push to it.

**Comment by schriftgestalt** [comment]
I'm still not through all your changes but it looks very good so far. Thanks. I added some small comments here and there.

**Comment by moyogo** [comment]
I’ll open a PR so we can start reviewing the branch.
